### PR TITLE
[Serializer] Introduce named serializers

### PR DESCRIPTION
**Pull request by HypeMC**
| Q             | A
| ------------- | ---
| Branch?       | 7.2
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        | -
| License       | MIT

The idea behind this PR is to allow configuring multiple serializer instances with different default contexts, name converters, and sets of normalizers and encoders. This is useful when your application is communicating with multiple APIs, each with different rules. Similar ideas have been mentioned [before](https://github.com/symfony/symfony/pull/35374#issuecomment-577177546).

```yaml
serializer:
  named_serializers:
    api1:
      name_converter: 'serializer.name_converter.camel_case_to_snake_case'
      default_context:
        enable_max_depth: true
    api2:
      default_context:
        enable_max_depth: false
```

The different serializers can be injected using named aliases:

```php
#[AsController]
class HomeController
{
    #[Route('/', name: 'app_home')]
    public function index(
        SerializerInterface $serializer,     // Default serializer
        SerializerInterface $api1Serializer, // api1 serializer
        #[Target('api2.serializer')]         // api2 serializer
        SerializerInterface $someName,
    ) {
        // ...
    }
}
```

Multiple normalizer/encoder instances with different arguments are created as child services of the default ones. ~~I also ensured that the same normalizer/encoder instances are reused between different serializers that have the same default context and name converter to minimize the number of child services created.~~

Custom normalizers/encoders can target specific serializers using the `serializer` tag attribute:

```yaml
get_set_method_normalizer:
    class: Symfony\Component\Serializer\Normalizer\GetSetMethodNormalizer
    autoconfigure: false # This is needed so that it's not included in the default serializer
    tags:
        # single serializer
        - serializer.normalizer: { serializer: 'api1' }
        # or multiple ones
        - serializer.normalizer: { serializer: [ 'api1', 'api2' ] }
        # use * to include the service in all serializers including the default one
        - serializer.normalizer: { serializer: '*' }
```

For BC reasons, not setting the `serializer` tag attribute is the same as setting it to the default one:

```yaml
    tags:
        - serializer.normalizer
        # same as
        - serializer.normalizer: { serializer: 'default' }
```

The profiler has been updated to support multiple serializer instances:

image

All normalizers/encoders are tagged with additional named serializer specific tags to help with debugging. To get the priority of normalizers/encoders for a certain serializer, use the `serializer.normalizer.<name>` or `serializer.encoder.<name>` tags:

```shell
$ bin/console debug:container --tag serializer.normalizer.default
Symfony Container Services Tagged with "serializer.normalizer.default" Tag
==========================================================================

 ------------------------------------------------- ---------- ------------------------------------------------------------------------------------------- 
  Service ID                                        priority   Class name                                                                                 
 ------------------------------------------------- ---------- ------------------------------------------------------------------------------------------- 
  serializer.denormalizer.unwrapping                1000       Symfony\Component\Serializer\Normalizer\UnwrappingDenormalizer                             
  serializer.normalizer.flatten_exception           -880       Symfony\Component\Messenger\Transport\Serialization\Normalizer\FlattenExceptionNormalizer  
  ...
  serializer.denormalizer.array                     -990       Symfony\Component\Serializer\Normalizer\ArrayDenormalizer                                  
  serializer.normalizer.object                      -1000      Symfony\Component\Serializer\Normalizer\ObjectNormalizer                                   
 ------------------------------------------------- ---------- ------------------------------------------------------------------------------------------- 

$ bin/console debug:container --tag serializer.normalizer.api1

Symfony Container Services Tagged with "serializer.normalizer.api1" Tag
=======================================================================

 ------------------------------------------------- ---------- ------------------------------------------------------------------------------------------- 
  Service ID                                        priority   Class name                                                                                 
 ------------------------------------------------- ---------- ------------------------------------------------------------------------------------------- 
  serializer.denormalizer.unwrapping                1000       Symfony\Component\Serializer\Normalizer\UnwrappingDenormalizer                             
  get_set_method_normalizer                                    Symfony\Component\Serializer\Normalizer\GetSetMethodNormalizer                             
  serializer.normalizer.flatten_exception           -880       Symfony\Component\Messenger\Transport\Serialization\Normalizer\FlattenExceptionNormalizer  
  ...
  serializer.denormalizer.array                     -990       Symfony\Component\Serializer\Normalizer\ArrayDenormalizer                                  
  serializer.normalizer.object                      -1000      Symfony\Component\Serializer\Normalizer\ObjectNormalizer                                   
 ------------------------------------------------- ---------- ------------------------------------------------------------------------------------------- 

$ bin/console debug:container --tag serializer.normalizer

Symfony Container Services Tagged with "serializer.normalizer" Tag
==================================================================

 ------------------------------------------------- ---------- ---------- ----------------- ------------------------------------------------------------------------------------------- 
  Service ID                                        standard   priority   serializer        Class name                                                                                 
 ------------------------------------------------- ---------- ---------- ----------------- ------------------------------------------------------------------------------------------- 
  serializer.denormalizer.unwrapping                1          1000                         Symfony\Component\Serializer\Normalizer\UnwrappingDenormalizer                             
  get_set_method_normalizer                                               ["api1","api2"]   Symfony\Component\Serializer\Normalizer\GetSetMethodNormalizer                             
  serializer.normalizer.flatten_exception           1          -880                         Symfony\Component\Messenger\Transport\Serialization\Normalizer\FlattenExceptionNormalizer  
  ...
  serializer.denormalizer.array                     1          -990                         Symfony\Component\Serializer\Normalizer\ArrayDenormalizer                                  
  serializer.normalizer.object                      1          -1000                        Symfony\Component\Serializer\Normalizer\ObjectNormalizer                                   
 ------------------------------------------------- ---------- ---------- ----------------- ------------------------------------------------------------------------------------------- 
```

Since Symfony comes with some pre-registered normalizers and encoders, I added options to exclude those in case someone wants to use only custom ones:

```yaml
serializer:
  named_serializers:
    api1:
      include_built_in_normalizers: false
      include_built_in_encoders: true

      name_converter: 'serializer.name_converter.camel_case_to_snake_case'
      default_context:
        enable_max_depth: true
```

TBH, I have doubts about the usefulness of this, please let me know your thoughts.

I've split the PR into two commits to ease reviewing:

- the first commit only rearranges the `SerializerPass` without adding any features
- the second commit implements the feature

### CODE REVIEW

**Comment by dunglas** [comment]
Good work! This will help improve performance and reduce the complexity of projects needing multiple serializers to deal with different contexts (ex: serializing data exposed through a public API and deserializing data coming from a third-party service). I'm +1 to merge this.

> `autoconfigure: false # This is needed so that it's not included in the default serializer`

Maybe could we tweak the autoconfigurator to automatically exclude definitions already having an explicit `serializer` tag? This will improve the DX a bit.

File: src/Symfony/Component/DependencyInjection/ContainerBuilder.php [hunk 1]
```diff
@@ -888,6 +888,17 @@ public function register(string $id, ?string $class = null): Definition
         return $this->setDefinition($id, new Definition($class));
     }
 
+    /**
+     * Registers a service definition.
+     *
+     * This method allows for simple registration of service definition
+     * with a fluid interface.
+     */
+    public function registerChild(string $id, string $parent): Definition
```
Comment (dunglas) [inline src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: This change is unrelated, could you propose it in another PR to keep this one focused?

**Review by dunglas — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/DependencyInjection/ContainerBuilder.php (hunk 1 above)
Comment (HypeMC) [reply src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: Done, see #58004

**Review by dunglas — APPROVED** [review]

**Review by mtarld — APPROVED** [review]
That's a really great addition!

Just a quick question, why mixing up `standard` and `default` as name describing default stuff? Shouldn't we stick to `default` only?

File: src/Symfony/Component/DependencyInjection/ContainerBuilder.php (hunk 1 above)
Comment (fabpot) [reply src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: #58004 has been merged now, @HypeMC Can you rebase this one when you have time?

**Comment by stof** [comment]
Reusing the same normalizer instances is a very bad idea as it would break for any normalizer implementing any of the `*Aware` interfaces of the component (as instantiating the second serializer would replace the instance that those shared normalizers are aware of)

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/serializer.html.twig [hunk 2]
```diff
@@ -115,20 +115,33 @@
...
             <div class="sf-tabs">
-                {{ _self.render_serialize_tab(collector.data, true) }}
-                {{ _self.render_serialize_tab(collector.data, false) }}
-
-                {{ _self.render_normalize_tab(collector.data, true) }}
-                {{ _self.render_normalize_tab(collector.data, false) }}
-
-                {{ _self.render_encode_tab(collector.data, true) }}
-                {{ _self.render_encode_tab(collector.data, false) }}
+                {% for serializer in collector.serializerNames %}
```
Comment (stof) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/serializer.html.twig:118]: WebProfilerBundle needs to have a conflict rule ensure it is not used with an older version of the collector

File: src/Symfony/Component/Serializer/Debug/TraceableNormalizer.php [hunk 3]
```diff
@@ -31,6 +31,7 @@ class TraceableNormalizer implements NormalizerInterface, DenormalizerInterface,
     public function __construct(
         private NormalizerInterface|DenormalizerInterface $normalizer,
         private SerializerDataCollector $dataCollector,
+        private readonly string $serializerName,
```
Comment (stof) [inline src/Symfony/Component/Serializer/Debug/TraceableNormalizer.php]: Adding a new required constructor argument is a BC break

File: src/Symfony/Component/Serializer/Debug/TraceableSerializer.php [hunk 4]
```diff
@@ -32,6 +32,7 @@ class TraceableSerializer implements SerializerInterface, NormalizerInterface, D
     public function __construct(
         private SerializerInterface&NormalizerInterface&DenormalizerInterface&EncoderInterface&DecoderInterface $serializer,
         private SerializerDataCollector $dataCollector,
+        private readonly string $serializerName,
```
Comment (stof) [inline src/Symfony/Component/Serializer/Debug/TraceableSerializer.php]: Adding a new constructor argument is a BC break. And marking those traceable classes as `@internal` is bad, because FrameworkBundle is the one instantiating them, which relies on BC (so that updating `symfony/serializer` does not break a project still using `framework-bundle` 7.1). They should not have been marked as `@internal`

**Review by stof — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/DependencyInjection/ContainerBuilder.php (hunk 1 above)
Comment (HypeMC) [reply src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: @fabpot Done

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/serializer.html.twig (hunk 2 above)
Comment (HypeMC) [reply src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/serializer.html.twig:118]: Done

File: src/Symfony/Component/Serializer/Debug/TraceableNormalizer.php (hunk 3 above)
Comment (HypeMC) [reply src/Symfony/Component/Serializer/Debug/TraceableNormalizer.php]: Fixed, I add a default value.

File: src/Symfony/Component/Serializer/Debug/TraceableSerializer.php (hunk 4 above)
Comment (HypeMC) [reply src/Symfony/Component/Serializer/Debug/TraceableSerializer.php]: Makes sense, fixed. I've removed the `@internal` tags as well.

**Comment by dunglas** [comment]
Good catch @stof, as usual. Having extra instances will be cleaner and will likely have a negligible overhead anyway, so let's simplify!

**Comment by HypeMC** [comment]
> > `autoconfigure: false # This is needed so that it's not included in the default serializer`
> 
> Maybe could we tweak the autoconfigurator to automatically exclude definitions already having an explicit `serializer` tag? This will improve the DX a bit.

@dunglas I agree, but this issue isn't limited to just this tag. I've encountered it many times. Whenever a service is manually tagged, and it has an autoconfigured tag, multiple tags are added to the definition.


> Just a quick question, why mixing up `standard` and `default` as name describing default stuff? Shouldn't we stick to `default` only?

@mtarld It didn't make sense to use the same name for two different concepts. `standard` normalizers/encoders aren't linked to the `default` serializer. They are services added to all serializer instances, not just the default one, unless the `include_standard_*` options are set to false.


> Reusing the same normalizer instances is a very bad idea as it would break for any normalizer implementing any of the `*Aware` interfaces of the component (as instantiating the second serializer would replace the instance that those shared normalizers are aware of)


@stof Tnx, I wasn't aware of those interfaces. I thought it would be a good idea to minimize the number of child services, but knowing this now, it's best to register new ones for each serializer instance. Fixed in https://github.com/symfony/symfony/compare/764b46f4ceee50b8f4148a7daab7f122d058b1c0..7b7d33f45e0d355aac0741de216b96e1853a151b

File: src/Symfony/Component/Serializer/Debug/TraceableNormalizer.php [hunk 5]
```diff
@@ -23,14 +23,13 @@
  * Collects some data about normalization.
  *
  * @author Mathias Arlaud <mathias.arlaud@gmail.com>
- *
- * @internal
```
Comment (chalasr) [inline src/Symfony/Component/Serializer/Debug/TraceableNormalizer.php]: should be in its own PR for 6.4 + replaced by `@final`

File: src/Symfony/Component/Serializer/Debug/TraceableSerializer.php [hunk 6]
```diff
@@ -22,8 +22,6 @@
  * Collects some data about serialization.
  *
  * @author Mathias Arlaud <mathias.arlaud@gmail.com>
- *
- * @internal
```
Comment (chalasr) [inline src/Symfony/Component/Serializer/Debug/TraceableSerializer.php]: same

**Review by chalasr — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Serializer/Debug/TraceableNormalizer.php (hunk 5 above)
Comment (HypeMC) [reply src/Symfony/Component/Serializer/Debug/TraceableNormalizer.php]: Done, see #58013

**Comment by chalasr** [comment]
> @mtarld It didn't make sense to use the same name for two different concepts. standard normalizers/encoders aren't linked to the default serializer. They are services added to all serializer instances, not just the default one, unless the include_standard_* options are set to false.

Not sure about the `standard` term either. Those are *built-in* encoders/normalizers as per the [docs](https://symfony.com/doc/current/components/serializer.html#built-in-normalizers). The term `default` looks also more appropriate in case it makes things simpler and I don't think the risk of ambiguity between the default serializer and those is worth caring - but `built-in` is certainly the best choice.

**Comment by HypeMC** [comment]
> > @mtarld It didn't make sense to use the same name for two different concepts. standard normalizers/encoders aren't linked to the default serializer. They are services added to all serializer instances, not just the default one, unless the include_standard_* options are set to false.
> 
> Not sure about the `standard` term either. Those are _built-in_ encoders/normalizers as per the [docs](https://symfony.com/doc/current/components/serializer.html#built-in-normalizers). The term `default` looks also more appropriate in case it makes things simpler and I don't think the risk of ambiguity between the default serializer and those is worth caring - but `built-in` is certainly the best choice.

@chalasr I replaced `standard` with `built-in`. I still think it makes sense to not use `default` as it might cause confusion, but if needed, I can change it to that instead.

https://github.com/symfony/symfony/compare/7b7d33f45e0d355aac0741de216b96e1853a151b..9e76b020d77ccb4d42ef0c00598221598e1e28e5

File: src/Symfony/Component/Serializer/DependencyInjection/SerializerPass.php [hunk 7]
```diff
@@ -31,44 +32,177 @@ class SerializerPass implements CompilerPassInterface
...
+    {
+        $defaultSerializerNameConverter = $container->hasParameter('.serializer.name_converter')
+            ? $container->getParameter('.serializer.name_converter') : null;
+
+        foreach ($container->getParameter('.serializer.named_serializers') as $serializerName => $config) {
+            $config += ['default_context' => [], 'name_converter' => null];
+            $serializerId = 'serializer.'.$serializerName;
+
+            if (!$normalizers = $this->findAndSortTaggedServices('serializer.normalizer.'.$serializerName, $container)) {
+                throw new RuntimeException(\sprintf('You must tag at least one service as "serializer.normalizer" with the name "%s" to use the "%s" service.', $serializerName, $serializerId));
```
Comment (chalasr) [inline src/Symfony/Component/Serializer/DependencyInjection/SerializerPass.php]: The message is not very actionable given `with the name`. We are rather looking for a `serializer` tag attribute right?
 
```suggestion
                throw new RuntimeException(\sprintf('Named serializer "%1$s" requires to have at least one normalizer registered. Tag them as "serializer.normalizer" with "serializer" attribute set to "%1$s".', $serializerName));
```
Maybe?

Comment (HypeMC) [reply src/Symfony/Component/Serializer/DependencyInjection/SerializerPass.php]: Done https://github.com/symfony/symfony/compare/9e76b020d77ccb4d42ef0c00598221598e1e28e5..7d850a081dbe00c0cb3050c12b53e44376aa9887

**Comment by chalasr** [comment]
Rebase needed after https://github.com/symfony/symfony/pull/58013

**Comment by HypeMC** [comment]
> Rebase needed after #58013

@chalasr Done

**Review by chalasr — APPROVED** [review]
Nice work

**Comment by chalasr** [comment]
fabbot's suggestion looks legit, can you apply it?

**Comment by HypeMC** [comment]
> fabbot's suggestion looks legit, can you apply it?

@chalasr Done, https://github.com/symfony/symfony/compare/d9ceb8521073610eff9d338eea2a5558d3edbbc5..03534988f5e9cef66798ab9b36d38cdf3770e671

File: src/Symfony/Component/Serializer/DependencyInjection/SerializerPass.php [hunk 8]
```diff
@@ -31,44 +32,177 @@ class SerializerPass implements CompilerPassInterface
...
+            foreach ($tags as $tag) {
+                $names = (array) ($tag['serializer'] ?? []);
+
+                if (!$names) {
+                    $names = ['default'];
+                } elseif (\in_array('*', $names, true)) {
+                    $names = array_unique(['default', ...$serializerNames]);
+                }
+
+                if ($tag['built-in'] ?? false) {
```
Comment (fabpot) [inline src/Symfony/Component/Serializer/DependencyInjection/SerializerPass.php]: To be consistent with the code base (like `extended_type` for Forms), I think it should be:

```suggestion
                if ($tag['built_in'] ?? false) {
```

**Review by fabpot — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Serializer/DependencyInjection/SerializerPass.php (hunk 8 above)
Comment (HypeMC) [reply src/Symfony/Component/Serializer/DependencyInjection/SerializerPass.php]: @fabpot Done, https://github.com/symfony/symfony/compare/aee8227eddd18e78b95b177ad5ea50d38c2c153a..ca834e5375c55edc7960417114ca3d02bd9f9845

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @HypeMC.
